(RK-291) Cache Forge module releases.

andersonmills opened the pull request:
This commit adds caching to the forge module releases. This is external
to environments, so will significantly increase the speed of environment
deploys across environments using the same forge module releases.

andersonmills committed:
(RK-291) Cache Forge module releases.

This commit adds caching to the forge module releases. This is external
to environments, so will significantly increase the speed of environment
deploys across environments using the same forge module releases.

haus commented on lib/r10k/forge/module_release.rb:
```diff
@@ -13,6 +13,7 @@ class ModuleRelease
 
       def_setting_attr :proxy
       def_setting_attr :baseurl
+      def_setting_attr :cache_root, File.expand_path(ENV['HOME'] ? '~/.r10k/cache': '/root/.r10k/cache')
```
If home isn't set, why do we choose `/root` here and do we think it will be writable? If running as a service user or via a script that scrubs env vars, it seems like $HOME probably won't be set and the user won't be able to write to `/root`.

haus replied:
But it seems like there are other settings like that, so maybe it's fine.

andersonmills replied:
Here, I was following the original git cache settings. Do you want to move this to /tmp if no $HOME?

haus replied:
i don't have context on why the other cache was in root to begin with, so it's probably fine, just struck me as odd.

andersonmills replied:
Agreed, that does seem weird. 

It is documented, though. :)

> The cachedir setting defaults to ~/.r10k. If the HOME environment variable is unset r10k will assume that r10k is being run with the Puppet prerun_command setting and will set the cachedir default to /root/.r10k.

haus commented on spec/unit/forge/module_release_spec.rb:
```diff
@@ -9,33 +9,57 @@
+      it "downloads the module from the forge into `download_path`" do
+        expect(tarball_cache_path).to receive(:exist?).and_return(false)
+        expect(subject.forge_release).to receive(:download).with(download_path)
+        allow(FileUtils).to receive(:mkdir_p).with(anything())
```
Do we want to be more specific about which directories are going to be created here?

andersonmills replied:
Will use instance_variable_set to set that, otherwise it's controlled by the presence/absence of r10k.yaml.

haus commented on lib/r10k/git/cache.rb:
```diff
@@ -16,7 +16,7 @@ class R10K::Git::Cache
   include R10K::Settings::Mixin
 
-  def_setting_attr :cache_root, File.expand_path(ENV['HOME'] ? '~/.r10k/git': '/root/.r10k/git')
+  def_setting_attr :cache_root, File.expand_path(ENV['HOME'] ? '~/.r10k/cache': '/root/.r10k/cache')
```
Why change the default location for the git cache?

andersonmills replied:
I changed this because originally, there was only the git cache. Now, we're adding the forge module cache to it, and in my opinion, they should be using the same root.

haus replied:
does the git cache just go immediately under the directory? i'm curious about how tarballs and git caches will be laid out together here.

do you have tree output or something from a cache dir with a mix of the two?

andersonmills replied:
Ah. forge tarball caching happens in a directory named for the module, ex branan-eight_hundred-8.0.0/, and git module and control caching happens in a directory named for the repo, ex. git@github.com-andersonmills-r10k-testing.git/.

haus replied:
okay, cool. that makes sense i just wanted to make sure they wouldn't cross paths.

ajroetker replied:
Should we document that they need to cleanup the old cache? Or emit a warning for them that their old cache will need a delete?

andersonmills replied:
Yes and no. In r10k documentation, we don't talk about the cache in a way which suggests the user might do anything with it. It might be worth noting this in the CHANGELOG, for those interested in that level of detail? I don't have another location which seems appropriate for a mid-version cache clearing.

As for code-manager/PE, we always specify the cache location.

ajroetker replied:
The cache won't get cleaned up for them then @andersonmills, how big is that cache?

andersonmills replied:
Nowadays, that cache is as big as the data they want to import for the module. We have one customer who had 10G in a cache.

andersonmills replied:
So, documenting where the cache is AND was seems like a good idea. Letting them know that they can blow the cache away, and the result will simply be that it is rebuilt also seems important.

haus reviewed (commented):
Looks pretty good to me. I had a few comments/questions, mainly from unfamiliarity with r10k.

ajroetker commented on lib/r10k/forge/module_release.rb:
```diff
@@ -41,9 +46,14 @@ def initialize(full_name, version)
-        @download_path = Pathname.new(Dir.mktmpdir) + (@forge_release.slug + '.tar.gz')
+        tarball_name = @forge_release.slug + '.tar.gz'
+        @download_path = Pathname.new(Dir.mktmpdir) + (tarball_name)
+        @tarball_cache_root = Pathname.new(settings[:cache_root]) + (@forge_release.slug + "/tarball/")
```
Does this variable need an `attr_accessor` above too? (as with https://github.com/puppetlabs/r10k/pull/729/files#diff-41ce558d2e9e83cf146b0fad6c9811d6R33)

andersonmills replied:
Need? No. It would make testing in the spec test a bit more clean, though.

andersonmills replied:
This did make me do a search for that variable, and made me realize I missed changing some comments/documentation.

ajroetker commented on lib/r10k/forge/module_release.rb:
```diff
@@ -65,34 +75,47 @@ def install(target_dir)
-        logger.debug1 "Downloading #{@forge_release.slug} from #{PuppetForge::Release.conn.url_prefix} to #{@download_path}"
-        @forge_release.download(download_path)
+        if @tarball_cache_path.exist?
+          logger.debug1 "Cached copy of #{@forge_release.slug} tarball found"
```
Nit pick, could you change this to be an action like `Using cached copy of tarball for ..`

andersonmills replied:
Good point.

ajroetker commented on lib/r10k/forge/module_release.rb:
```diff
@@ -65,34 +75,47 @@ def install(target_dir)
+        if @tarball_cache_path.exist?
+          logger.debug1 "Cached copy of #{@forge_release.slug} tarball found"
+        else
+          logger.debug1 "Downloading #{@forge_release.slug} from #{PuppetForge::Release.conn.url_prefix} to #{@download_path}"
```
Why don't we just down directly to the cache directory?

ajroetker replied:
Actually, why do we need download_path at all anymore? Can they turn off cacheing?

andersonmills replied:
This still exists so there's a location that it's downloaded to, which takes time, and then it's mv'd to the cache directory, which is effectively atomic. It minimizes the chance of getting partial downloads stuck somewhere.

andersonmills replied:
They are currently not able to turn off caching.

ajroetker commented on lib/r10k/forge/module_release.rb:
```diff
@@ -105,7 +128,6 @@ def unpack(target_dir)
       # Remove all files created while downloading and unpacking the module.
       def cleanup
         cleanup_unpack_path
-        cleanup_download_path
```
Why aren't we calling this function anymore?

andersonmills replied:
Because putting this in the cache mv's the tarball now instead of unpacking from that location and leaving it there.

andersonmills replied:
This could be added back for the case of things getting stopped halfway through.

ajroetker reviewed (changes requested):
Should we have an option that bypasses the cache?
Should we have a command to cleanup the cache for them? (i.e. remove outdated modules that aren't in their enviroments?)

andersonmills commented on lib/r10k/forge/module_release.rb:
```diff
@@ -65,34 +79,47 @@ def install(target_dir)
       def verify
-        logger.debug1 "Verifying that #{download_path} matches checksum #{@forge_release.file_md5}"
-        @forge_release.verify(download_path)
+        logger.debug1 "Verifying that #{@tarball_cache_path} matches checksum #{@forge_release.file_md5}"
```
This should download the md5 and store it next to the tarball, and verify using that.

andersonmills committed:
(RK-291) Add storage of local tarball md5

This commit stores the md5 from the forge in the cache for later reuse.
On checksum mismatch, both cached files are deleted, forcing a reload of
both from the forge.

ajroetker commented on spec/unit/forge/module_release_spec.rb:
```diff
@@ -7,35 +7,110 @@
+      expect(subject).to receive(:cleanup_cached_tarball_path)
+      expect { subject.verify_from_forge(md5_of_tarball) }.to raise_error(PuppetForge::V3::Release::ChecksumMismatch)
+    end
+    #it "verifies the module checksum based on the Forge file checksum" do
```
Should this be commented out still?

andersonmills replied:
Ah. Oops. That should be completely removed. That test is superseded by the "write the md5 to file when the checksums match" test. Removing.
